p2p/protocol/identify/id.go

```diff
@@ -333,6 +334,7 @@ func (ids *idService) IdentifyWait(c network.Conn) <-chan struct{} {
 			if err := ids.identifyConn(c); err != nil {
```

Comment:
Closing the channel seems a bit dangerous to misuse, as it invites users to receive from the channel multiple times. Only in the first case they'll get the error, after that they'll get a `nil`.

Would it be safe to send a `nil` on the channel when no error occurs (without closing)?

Reply:
I'm not sure I follow the reasoning. What's wrong with closing the channel? We can send a nil in case of no error and then close (which is about the same as closing without sending a nil first). 

If a user is receiving multiple times without a closed channel they'll have a waiting recv that will never get completed, right? That seems like a caller bug.

Reply:
> What's wrong with closing the channel? We can send a nil in case of no error and then close (which is about the same as closing without sending a nil first).

That's correct. The problem is the error case, where the error is only received once, and after that it _looks_ as if no error occurred (as a receiving from a closed channel returns the zero value).

Not closing would be safer: The API contract would tell users that they can only expect to receive from this channel once.

Reply:
A user could check if the chan is closed with `x, ok := <-c`.

Maybe what we're really saying here is that we wish there was a way to encode that "this channel will return at most one error and then close". And because we can't we're forced to pick two options:

1. Close the channel, and maybe cause a caller to recv multiple times and get nil.
2. Don't close the channel, and maybe cause a caller to block forever on a second recv that will never come.

I think neither maybe case is worth worrying about, and I think you'd prefer the 2nd option if we had to worry about it.

Is that correct? I don't have a strong preference here, so happy to do 2. I just wanted to understand your reasoning.

Reply:
Because of the above later callers may not get the error. I need to rethink this